File: src/workload.py
import random
import sys

import numpy as np

if len(sys.argv) != 4:
    print("the script must be called with the probabilities ")

    raise Exception

weight_name = sys.argv[1]
temporal = float(sys.argv[2])
error = float(sys.argv[3])

n_vertices = int(input())
n_messages = int(input())

n_pairs = ((n_vertices - 1) * n_vertices) // 2
weights = {}

so_far = 0
with open(f"weights/{weight_name}") as fo:
    for line in fo.readlines():
        src, dst, weight = line.split(",")
        src = int(src)
        dst = int(dst)
        weight = int(weight)
        so_far += weight

        weights[(min(src, dst), max(src, dst))] = weight

n_tests = 30
for test in range(n_tests):
    og_probs = []
    pairs = []
    for src in range(n_vertices):
        for dst in range(src + 1, n_vertices):
            pairs.append((len(pairs), src, dst))
            og_probs.append(
                max(0, weights.get((src, dst), 1) +
                np.random.choice([error, -error]))
            )

    og_probs = np.array(og_probs) / np.sum(og_probs)
    cur_probs = og_probs.copy()
    seq = []

    same_as_last = 0
    last = -1

    for _ in range(n_messages):
        chosen, src, dst = random.choices(pairs, weights=cur_probs)[0]

        cur_probs = og_probs.copy() * (1 - temporal)
        cur_probs[chosen] += temporal

        if chosen == last:
            same_as_last += 1

        last = chosen
        seq.append((src, dst))

    with open(f"input/{weight_name}-{test}.txt", "w") as fo:
        fo.write(f"{n_vertices} {n_messages}\n")
        for src, dst in seq:
            fo.write(f"{src} {dst}\n")

File: src/workload.cc
#include <map>
#include <fstream>
#include <chrono>
#include <string>
#include <random>
#include <iostream>
#include <stdlib.h>


int main (int argc, char* argv[]) {
    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
    std::mt19937 gen(seed);

    if (argc < 4) {
        std::cout << "the script must be called with the probabilities" << std::endl;
        throw;

    }

    std::string weightName = argv[1];
    double temporalProb = atof(argv[2]);
    double error = atof(argv[3]);

    int nVertices, nMessages;
    std::map<std::pair<int,int>, int> weights;
    std::cin >> nVertices >> nMessages;

    std::ifstream weightFile("weights/" + weightName);

    int src, dst, weight;
    while (weightFile >> src >> dst >> weight) {
        weights[{std::min(src, dst), std::max(src, dst)}] = weight;

    }

    std::vector<std::pair<int, int>> pairs;
    for (int src = 0; src < nVertices; src++) {
        for (int dst = src + 1; dst < nVertices; dst++) {
            pairs.push_back({src, dst});

        }
    }

    for (int testIdx = 0; testIdx < 30; testIdx++) {
        double sum = 0;
        std::vector<double> ogProb, roundProb;
        for (int src = 0; src < nVertices; src++) {
            for (int dst = src + 1; dst < nVertices; dst++) {
                auto wIt = weights.find({src, dst});

                std::uniform_real_distribution<> errorDis(
                    -error, error
                );

                double weight = (
                    wIt == weights.end() ? 1.0 : wIt->second
                ) + errorDis(gen);

                ogProb.push_back(weight);

                sum += weight;
            }
        }

        for (double& w: ogProb)
            w /= sum;

        roundProb = ogProb;
        std::ofstream tFile(
            "input/" + weightName + "-" + std::to_string(testIdx) + ".txt"
        );

        tFile << nVertices << " " << nMessages << std::endl;

        int lastIdx = -1;
        for (int mIdx = 0; mIdx < nMessages; mIdx++) {
            std::discrete_distribution<int> pDis(
                std::begin(roundProb), std::end(roundProb)
            );

            int nextIdx = pDis(gen);
            tFile << pairs[nextIdx].first << " " << pairs[nextIdx].second << std::endl;

            if (nextIdx != lastIdx) {
                roundProb = ogProb;
                for (int pairIdx = 0; pairIdx < roundProb.size(); pairIdx++) {
                    roundProb[pairIdx] *= (1.0 - temporalProb);
                    if (pairIdx == nextIdx) {
                        roundProb[pairIdx] += temporalProb;

                    }
                }
            }

        }
    }
}